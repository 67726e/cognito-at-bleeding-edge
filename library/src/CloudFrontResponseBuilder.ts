import { CloudFrontRequestResult } from 'aws-lambda';

export interface CloudFrontResponseBuilder {
  buildCognitoRedirectResponse(location: string): CloudFrontRequestResult;
  buildCookieRedirectResponse(location: string, cookies: string[]): CloudFrontRequestResult;
}

export class DefaultCloudFrontResponseBuilder implements CloudFrontResponseBuilder {
  buildCognitoRedirectResponse(location: string): CloudFrontRequestResult {
    return {
      status: '302',
      headers: {
        'location': [{
          key: 'Location',
          value: location,
        }],
        'cache-control': [{
          key: 'Cache-Control',
          value: 'no-cache, no-store, max-age=0, must-revalidate',
        }],
        'pragma': [{
          key: 'Pragma',
          value: 'no-cache',
        }],
      },
    };
  }

  buildCookieRedirectResponse(location: string, cookies: string[]): CloudFrontRequestResult {
    return {
      status: '302' ,
      headers: {
        'location': [{
          key: 'Location',
          value: location,
        }],
        'cache-control': [{
          key: 'Cache-Control',
          value: 'no-cache, no-store, max-age=0, must-revalidate',
        }],
        'pragma': [{
          key: 'Pragma',
          value: 'no-cache',
        }],
        'set-cookie': cookies.map((cookie) => ({ key: 'Set-Cookie', value: cookie })),
      },
    };
  }
}
