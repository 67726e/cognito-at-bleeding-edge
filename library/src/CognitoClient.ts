import axios from 'axios';
import { stringify } from 'querystring';

import { Logger } from './Logger';
import { CognitoConfiguration, Tokens } from './types';

export interface CognitoClient {
  postAuthorizationCodeExchange(redirectURI: string, code: string): Promise<Tokens>;
  postRefreshTokenExchange(redirectURI: string, refreshToken: string): Promise<any>;
}

export class DefaultCognitoClient implements CognitoClient {
  constructor(
    private readonly cognitoConfiguration: CognitoConfiguration,
    private readonly logger: Logger,
  ) {}

  /**
   * Exchange Authorization Code for Access, ID, Refresh Tokens
   * @param  {String} redirectURI Redirection URI.
   * @param  {String} code        Authorization Code.
   * @return {Promise} Authorized User Tokens.
   */
  postAuthorizationCodeExchange(redirectURI: string, code: string): Promise<Tokens> {
    // TODO: Document Me! - Why Optional `userPoolAppStream`
    const authorization = this.cognitoConfiguration.userPoolAppSecret && Buffer.from(`${this.cognitoConfiguration.userPoolAppId}:${this.cognitoConfiguration.userPoolAppSecret}`).toString('base64');

    const request = {
      url: `https://${this.cognitoConfiguration.userPoolDomain}/oauth2/token`,
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
        ...(authorization && {'Authorization': `Basic ${authorization}`}),
      },
      data: stringify({
        client_id:	this.cognitoConfiguration.userPoolAppId,
        code:	code,
        grant_type:	'authorization_code',
        redirect_uri:	redirectURI,
      }),
    } as const;

    this.logger.debug({ msg: 'Fetching Tokens for Authorization Code', request, code });

    return axios.request(request)
      .then((response) => {
        this.logger.debug({ msg: 'Successfully Fetched Tokens for Authorization Code', tokens: response.data });

        // TODO: Validate / Assert Type!
        return {
          accessToken: response.data.access_token,
          idToken: response.data.id_token,
          refreshToken: response.data.refresh_token,
        };
      })
      .catch((error) => {
        this.logger.error({ msg: 'Failure Fetching Tokens for Authorization Code', request, code });

        throw error;
      });
  }

  /**
   * Exchange Refresh Token for Access, ID Tokens
   * @param  {String} redirectURI Redirection URI.
   * @param  {String} refreshToken Refresh Token.
   * @return {Promise<Tokens>} Refreshed User Tokens.
   */
  postRefreshTokenExchange(redirectURI: string, refreshToken: string) {
    // TODO: Document Me! - Why Optional `userPoolAppStream`
    const authorization = this.cognitoConfiguration.userPoolAppSecret && Buffer.from(`${this.cognitoConfiguration.userPoolAppId}:${this.cognitoConfiguration.userPoolAppSecret}`).toString('base64');

    const request = {
      url: `https://${this.cognitoConfiguration.userPoolDomain}/oauth2/token`,
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
        ...(authorization && {'Authorization': `Basic ${authorization}`}),
      },
      data: stringify({
        client_id: this.cognitoConfiguration.userPoolAppId,
        refresh_token: refreshToken,
        grant_type: 'refresh_token',
        redirect_uri: redirectURI,
      }),
    } as const;

    this.logger.debug({ msg: 'Fetching Tokens for Refresh Token', request, refreshToken });

    return axios.request(request)
      .then((response) => {
        this.logger.debug({ msg: 'Successfully Fetched Tokens for Refresh Token', tokens: response.data });
        
        // TODO: Validate / Assert Type!
        return {
          accessToken: response.data.access_token,
          idToken: response.data.id_token,
          // TODO: Add `refreshToken` ???
        };
      })
      .catch((error) => {
        this.logger.error({ msg: 'Failure Fetching Tokens for Refresh Token', request, refreshToken });
        
        throw error;
      });
  }
}
