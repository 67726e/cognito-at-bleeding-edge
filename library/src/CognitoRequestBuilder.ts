import { CognitoConfiguration } from './types';

export interface CognitoRequestBuilder {
  buildAuthorizeRequest(redirectURI: string, redirectPath: string): string;
}

export class DefaultCognitoRequestBuilder implements CognitoRequestBuilder {
  constructor(
    private readonly cognitoConfiguration: CognitoConfiguration,
  ) {}

  buildAuthorizeRequest(redirectURI: string, redirectPath: string): string {
    const authorizeUrl = `https://${this.cognitoConfiguration.userPoolDomain}/authorize?` +
      `client_id=${this.cognitoConfiguration.userPoolAppId}&` + 
      `redirect_uri=${redirectURI}&` +
      'response_type=code&' +
      `state=${redirectPath}`;

    return authorizeUrl;
  }
}
