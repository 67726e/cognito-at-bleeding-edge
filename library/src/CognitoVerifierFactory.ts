import { CognitoJwtVerifier, CognitoJwtVerifierProperties, CognitoJwtVerifierSingleUserPool } from 'aws-jwt-verify/cognito-verifier';
import { CognitoConfiguration } from './types';

export interface CognitoTokenVerifierFactory {
  createAccessTokenVerifier(): CognitoJwtVerifierSingleUserPool<CognitoJwtVerifierProperties>;
  createIdTokenVerifier(): CognitoJwtVerifierSingleUserPool<CognitoJwtVerifierProperties>;
}

export class DefaultCognitoTokenVerifierFactory implements CognitoTokenVerifierFactory {
  constructor(
    private readonly cognitoConfiguration: CognitoConfiguration,
  ) {}

  createAccessTokenVerifier() {
    return CognitoJwtVerifier.create({
      userPoolId: this.cognitoConfiguration.userPoolId,
      clientId: this.cognitoConfiguration.userPoolAppId,
      tokenUse: 'access',
    } as CognitoJwtVerifierProperties);
  }

  createIdTokenVerifier() {
    return CognitoJwtVerifier.create({
      userPoolId: this.cognitoConfiguration.userPoolId,
      clientId: this.cognitoConfiguration.userPoolAppId,
      tokenUse: 'id',
    } as CognitoJwtVerifierProperties);
  }
}
