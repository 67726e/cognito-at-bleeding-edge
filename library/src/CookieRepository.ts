import { CookieAttributes, Cookies } from './cookie-at-bleeding-edge';
import { CookieConfiguration, CookieHeaders, Tokens } from './types';

// TODO: Refactor Further...
export interface CookieRepository {
  getAccessTokenCookie(): string;
  getIdTokenCookie(): string;
  getRefreshTokenCookie(): string;

  createCookies(domain: string, tokens: Tokens): string[];

  parseCookies(cookieHeaders: CookieHeaders): Tokens;
}

export class DefaultCookieRepository implements CookieRepository {
  constructor(
    private readonly cookieConfiguration: CookieConfiguration,
  ) {}

  getAccessTokenCookie = () => 'cloudfront.accessToken';
  getIdTokenCookie = () => 'cloudfront.idToken';
  getRefreshTokenCookie = () => 'cloudfront.refreshToken';

  createCookies = (domain: string, tokens: Tokens): string[] => {
    const cookieAttributes: CookieAttributes = {
      domain: this.cookieConfiguration.disableDomain ? undefined : domain,
      expires: new Date(Date.now() + this.cookieConfiguration.expirationDays * 864e+5),
      httpOnly: this.cookieConfiguration.httpOnly,
      path: this.cookieConfiguration.path,
      sameSite: this.cookieConfiguration.sameSite,
      secure: true,
    };

    // TODO: Remove Me! `(tokens.accessToken || '')`
    const accessToken = Cookies.serialize(`${this.getAccessTokenCookie()}`, tokens.accessToken || '', cookieAttributes);
    // TODO: Remove Me! `(tokens.idToken || '')`
    const idToken = Cookies.serialize(`${this.getIdTokenCookie()}`, tokens.idToken || '', cookieAttributes);

    // TODO: Review Me!
    if (tokens.refreshToken) {
      const refreshToken = Cookies.serialize(`${this.getRefreshTokenCookie()}`, tokens.refreshToken, cookieAttributes);

      return [
        accessToken,
        idToken,
        refreshToken,
      ];
    } else {
      return [
        accessToken,
        idToken,
      ];
    }
  }

  parseCookies = (cookieHeaders: CookieHeaders): Tokens => {
    const cookies = (cookieHeaders || []).flatMap((header) => Cookies.parse(header.value));

    const tokens: Tokens = {};

    for (const {name, value} of cookies) {
      tokens.accessToken = (name === this.getAccessTokenCookie()) ? value : tokens.accessToken;
      tokens.idToken = (name === this.getIdTokenCookie()) ? value : tokens.idToken;
      tokens.refreshToken = (name === this.getRefreshTokenCookie()) ? value : tokens.refreshToken;
    }

    return tokens;
  }
}