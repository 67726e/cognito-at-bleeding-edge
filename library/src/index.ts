export * from './Authenticator';
export * from './CloudFrontResponseBuilder';
export * from './CognitoClient';
export * from './CognitoRequestBuilder';
export * from './CognitoVerifierFactory';
export * from './cookie-at-bleeding-edge';
export * from './CookieRepository';
export * from './DefaultAuthenticator';
export * from './Logger';
export * from './types';