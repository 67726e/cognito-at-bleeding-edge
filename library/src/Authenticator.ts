import { CloudFrontRequest, CloudFrontRequestEvent, CloudFrontRequestResult } from 'aws-lambda';
import { parse } from 'querystring';

import { CloudFrontResponseBuilder, } from './CloudFrontResponseBuilder';
import { CognitoClient, } from './CognitoClient';
import { CognitoRequestBuilder, } from './CognitoRequestBuilder';
import { SameSite, SAME_SITE_VALUES } from './cookie-at-bleeding-edge';
import { CookieRepository, } from './CookieRepository';
import { Logger } from './Logger';
import { CognitoConfiguration, CookieHeaders, Tokens } from './types';

export type DefaultAuthenticatorCognitoConfiguration = CognitoConfiguration

export interface DefaultAuthenticatorCookieConfiguration {
  disableDomain?: boolean;
  expirationDays?: number;
  httpOnly?: boolean;
  sameSite?: SameSite;
  path?: string;
}

export interface DefaultAuthenticatorConfiguration {
  cognitoConfiguration: DefaultAuthenticatorCognitoConfiguration;
  cookieConfiguration?: DefaultAuthenticatorCookieConfiguration;
}

export type AuthenticatorResult = {
  actual: CloudFrontRequestResult,
  isAuthenticated: boolean,
};

export class Authenticator {
  constructor(
    private readonly cloudFrontResponseBuilder: CloudFrontResponseBuilder,
    private readonly cognitoClient: CognitoClient,
    private readonly cognitoRequestBuilder: CognitoRequestBuilder,
    private readonly cookieRepository: CookieRepository,
    private readonly logger: Logger,
    private readonly accessTokenVerifier: any,
    private readonly idTokenVerifier: any,
  ) {}

  /**
   * Extract value of the authentication token from the request cookies.
   * @param  {Array}  cookieHeaders 'Cookie' request headers.
   * @return {Tokens} Extracted id token or access token. Null if not found.
   */
  _getTokensFromCookie(cookieHeaders: CookieHeaders): Tokens {
    if (!cookieHeaders) {
      this.logger.debug("Cookies weren't present in the request");

      throw new Error("Cookies weren't present in the request");
    }

    this.logger.debug({ msg: 'Extracting authentication token from request cookie', cookieHeaders });

    const tokens: Tokens = this.cookieRepository.parseCookies(cookieHeaders);

    if (!tokens.accessToken && !tokens.idToken && !tokens.refreshToken) {
      this.logger.debug('Neither accessToken, idToken, nor refreshToken was present in request cookies');

      throw new Error('Neither accessToken, idToken, nor refreshToken was present in request cookies');
    }

    this.logger.debug({ msg: 'Found tokens in cookie', tokens });

    return tokens;
  }

  /**
   * Create a Lambda@Edge redirection response to set the tokens on the user's browser cookies.
   * @param  {Object} tokens   Cognito User Pool tokens.
   * @param  {String} domain   Website domain.
   * @param  {String} location Path to redirection.
   * @return Lambda@Edge response.
   */
  _createSetCookieRedirectResponse(tokens: Tokens, domain: string, location: string): CloudFrontRequestResult {
    const cookies = this.cookieRepository.createCookies(domain, tokens);

    const response: CloudFrontRequestResult = this.cloudFrontResponseBuilder.buildCookieRedirectResponse(location, cookies);

    this.logger.debug({ msg: 'Generated set-cookie response', response });

    return response;
  }

  /**
   * Get redirect to cognito userpool response
   * @param  {CloudFrontRequest}  request The original request
   * @param  {string}  redirectURI Redirection URI.
   * @return {CloudFrontRequestResult} Redirect response.
   */
  _createCognitoAuthorizationRedirectResponse(request: CloudFrontRequest, redirectURI: string): CloudFrontRequestResult {
    const redirectPath = (request.querystring && request.querystring !== '') ?
      request.uri + encodeURIComponent('?' + request.querystring) :
      request.uri;

    const userPoolUrl = this.cognitoRequestBuilder.buildAuthorizeRequest(redirectURI, redirectPath);

    this.logger.debug(`Redirecting user to Cognito User Pool URL ${userPoolUrl}`);

    return this.cloudFrontResponseBuilder.buildCognitoRedirectResponse(userPoolUrl);
  }

  /**
   * Handle Lambda@Edge event:
   *   * if authentication cookie is present and valid: forward the request
   *   * if authentication cookie is invalid, but refresh token is present: set cookies with refreshed tokens
   *   * if ?code=<grant code> is present: set cookies with new tokens
   *   * else redirect to the Cognito UserPool to authenticate the user
   * @param  {Object}  event Lambda@Edge event.
   * @return {Promise} CloudFront response.
   */
  async handle(event: CloudFrontRequestEvent): Promise<AuthenticatorResult> {
    this.logger.debug({ msg: 'Handling Lambda@Edge Event', event });

    const { request } = event.Records[0].cf;
    const requestParams = parse(request.querystring);
    const cloudFrontDomain = request.headers.host[0].value;
    const redirectURI = `https://${cloudFrontDomain}`;

    // 1. Get Token(s) from Cookie Header(s)
    //  Cookie(s) May or May Not Include: Access Token, ID Token, Refresh Token
    // 2. Verify Access Token, ID Token
    //  If Access Token and ID Token verify, return original request.
    //  If Access Token and or ID Token fail, next step 
    // 3. Fetch Token(s) from Refresh Token
    //  If Fetch Token(s) succeed, return redirect request w/ Access Token, ID Token as `Set-Cookie` Header
    //  If Fetch Token(s) fail, next step
    // 4. Get Authentication Code from Query Parameter(s), Exchange for Token(s)
    //  If Fetch Token(s) success, return redirect request w/ Access Token, ID Token as `Set-Cookie` Header
    //  If Fetch Token(s) fail, final failure
    // 5. Return Redirect to Cognito Authentication Pool

    try {
      // 1. Get Token(s) from Cookie Header(s)
      const tokens = this._getTokensFromCookie(request.headers.cookie);

      this.logger.debug({ msg: 'Verifying token...', tokens });

      try {
        // 2. Verify Access Token, ID Token
        const identity = await this.idTokenVerifier.verify(tokens.idToken);
        const access = await this.accessTokenVerifier.verify(tokens.accessToken);

        this.logger.info({ msg: 'Forwarding request', path: request.uri, identity, access });

        return { actual: request, isAuthenticated: true, };
      } catch (err) {
        // 3. Fetch Token(s) from Refresh Token
        if (tokens.refreshToken) {
          this.logger.debug({ msg: 'Verifying accessToken, idToken failed, verifying refresh token instead...', tokens, err });

          return await this.cognitoClient.postRefreshTokenExchange(redirectURI, tokens.refreshToken)
            .then(tokens => {
              const result = this._createSetCookieRedirectResponse(tokens, cloudFrontDomain, request.uri);

              return { actual: result, isAuthenticated: false, };
            });
        } else {
          throw err;
        }
      }
    } catch (err) {
      this.logger.debug("User isn't authenticated: %s", err);

      if (requestParams.code) {
        // 4. Get Authentication Code from Query Parameter(s), Exchange for Token(s)
        return await this.cognitoClient.postAuthorizationCodeExchange(redirectURI, requestParams.code as string)
          .then(tokens => {
            const result = this._createSetCookieRedirectResponse(tokens, cloudFrontDomain, requestParams.state as string);

            return { actual: result, isAuthenticated: false, };
          });
      } else {
        // 5. Return Redirect to Cognito Authentication Pool
        const result = this._createCognitoAuthorizationRedirectResponse(request, redirectURI);

        return { actual: result, isAuthenticated: false, };
      }
    }
  }
}
