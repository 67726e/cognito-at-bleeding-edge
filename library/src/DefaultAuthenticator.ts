import { CloudFrontRequestEvent, CloudFrontRequestResult } from "aws-lambda";
import { Authenticator, DefaultAuthenticatorConfiguration, DefaultAuthenticatorCookieConfiguration } from "./Authenticator";
import { DefaultCloudFrontResponseBuilder } from "./CloudFrontResponseBuilder";
import { DefaultCognitoClient } from "./CognitoClient";
import { DefaultCognitoRequestBuilder } from "./CognitoRequestBuilder";
import { CognitoTokenVerifierFactory, DefaultCognitoTokenVerifierFactory } from "./CognitoVerifierFactory";
import { SAME_SITE_VALUES } from "./cookie-at-bleeding-edge";
import { DefaultCookieRepository } from "./CookieRepository";
import { Logger, DefaultLogger } from "./Logger";
import { CognitoConfiguration, CookieConfiguration } from "./types";

export class DefaultAuthenticator {
    private authenticator;

    constructor(
        configuration: DefaultAuthenticatorConfiguration,
    ) {
        const cognitoConfiguration: CognitoConfiguration = configuration.cognitoConfiguration;
        const cookieConfiguration: CookieConfiguration = DefaultAuthenticator.buildCookieConfiguration(configuration.cookieConfiguration);

        const cognitoTokenVerifierFactory: CognitoTokenVerifierFactory = new DefaultCognitoTokenVerifierFactory(
            cognitoConfiguration,
        );

        const defaultLogger: Logger = new DefaultLogger();

        this.authenticator = new Authenticator(
            new DefaultCloudFrontResponseBuilder(),
            new DefaultCognitoClient(
                cognitoConfiguration,
                defaultLogger,
            ),
            new DefaultCognitoRequestBuilder(
                cognitoConfiguration,
            ),
            new DefaultCookieRepository(
                cookieConfiguration,
            ),
            defaultLogger,
            cognitoTokenVerifierFactory.createAccessTokenVerifier(),
            cognitoTokenVerifierFactory.createIdTokenVerifier(),
        );
    }

    static buildCookieConfiguration(cookieConfiguration?: DefaultAuthenticatorCookieConfiguration): CookieConfiguration {
        return {
            // TODO: Review Default!
            disableDomain: (cookieConfiguration && typeof cookieConfiguration.disableDomain === 'boolean') ? cookieConfiguration.disableDomain : false,
            // TODO: Review Default!
            expirationDays: (cookieConfiguration && typeof cookieConfiguration.expirationDays === 'number') ? cookieConfiguration.expirationDays : 365,
            // TODO: Review Default!
            httpOnly: (cookieConfiguration && typeof cookieConfiguration.httpOnly === 'boolean') ? cookieConfiguration.httpOnly : undefined,
            // We must intentionally set the `cookiePath` to `/`, providing access to all paths on the domain.
            // Otherwise, the path appears set by the browser(?) and results in duplicate cookies and a redirect loop.
            // See: https://github.com/awslabs/cognito-at-edge/issues/69
            path: (cookieConfiguration && typeof cookieConfiguration.path === 'string') ? cookieConfiguration.path : '/',
            // TODO: Review Default!
            sameSite: (cookieConfiguration && cookieConfiguration.sameSite !== undefined && SAME_SITE_VALUES.includes(cookieConfiguration.sameSite)) ? cookieConfiguration.sameSite : undefined,
        };
    }

    async handle(event: CloudFrontRequestEvent): Promise<CloudFrontRequestResult> {
        return this.authenticator.handle(event);
    }
}
