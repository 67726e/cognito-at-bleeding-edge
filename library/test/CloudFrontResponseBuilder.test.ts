
import { beforeEach, describe, expect, test } from '@jest/globals'

import { CloudFrontResponseBuilder, DefaultCloudFrontResponseBuilder } from '../src/CloudFrontResponseBuilder';

describe('DefaultCloudFrontResponseBuilder', () => {
    let builder: CloudFrontResponseBuilder;

    beforeEach(() => {
        builder = new DefaultCloudFrontResponseBuilder();
    });

    test('buildCognitoRedirectResponse', () => {
        const response =  builder.buildCognitoRedirectResponse('https://cloudfront-pool-domain.auth.us-east-1.amazoncognito.com/authorize?client_id=1hoa20u57ikp7nafi8ks287uba&redirect_uri=https://d20o2ten651kje.cloudfront.net&response_type=code&state=/        ');

        // https://cloudfront-pool-domain.auth.us-east-1.amazoncognito.com/authorize?client_id=1hoa20u57ikp7nafi8ks287uba&redirect_uri=https://d20o2ten651kje.cloudfront.net&response_type=code&state=/

        expect(response).toMatchSnapshot();
    });

    test('buildCookieRedirectResponse', () => {
        const response =  builder.buildCookieRedirectResponse('', [
            'cloudfront.accessToken=MOCK_ACCESS_TOKEN; Domain=d20o2ten651kje.cloudfront.net; Path=/; Expires=Mon, 21 Oct 2024 20:28:19 GMT; Secure',
            'cloudfront.idToken=MOCK_ID_TOKEN; Domain=d20o2ten651kje.cloudfront.net; Path=/; Expires=Mon, 21 Oct 2024 20:28:19 GMT; Secure',
            'cloudfront.refreshToken=MOCK_REFRESH_TOKEN; Domain=d20o2ten651kje.cloudfront.net; Path=/; Expires=Mon, 21 Oct 2024 20:28:19 GMT; Secure',
        ]);

        // 'set-cookie: cloudfront.accessToken=MOCK_ACCESS_TOKEN; Domain=d20o2ten651kje.cloudfront.net; Path=/; Expires=Mon, 21 Oct 2024 20:28:19 GMT; Secure',
        // 'set-cookie: cloudfront.idToken=MOCK_ID_TOKEN; Domain=d20o2ten651kje.cloudfront.net; Path=/; Expires=Mon, 21 Oct 2024 20:28:19 GMT; Secure',
        // 'set-cookie: cloudfront.refreshToken=MOCK_REFRESH_TOKEN; Domain=d20o2ten651kje.cloudfront.net; Path=/; Expires=Mon, 21 Oct 2024 20:28:19 GMT; Secure',
    
        expect(response).toMatchSnapshot();
    });
});
