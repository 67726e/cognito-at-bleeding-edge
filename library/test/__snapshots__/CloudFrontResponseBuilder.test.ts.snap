// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DefaultCloudFrontResponseBuilder buildCognitoRedirectResponse 1`] = `
{
  "headers": {
    "cache-control": [
      {
        "key": "Cache-Control",
        "value": "no-cache, no-store, max-age=0, must-revalidate",
      },
    ],
    "location": [
      {
        "key": "Location",
        "value": "https://cloudfront-pool-domain.auth.us-east-1.amazoncognito.com/authorize?client_id=1hoa20u57ikp7nafi8ks287uba&redirect_uri=https://d20o2ten651kje.cloudfront.net&response_type=code&state=/        ",
      },
    ],
    "pragma": [
      {
        "key": "Pragma",
        "value": "no-cache",
      },
    ],
  },
  "status": "302",
}
`;

exports[`DefaultCloudFrontResponseBuilder buildCookieRedirectResponse 1`] = `
{
  "headers": {
    "cache-control": [
      {
        "key": "Cache-Control",
        "value": "no-cache, no-store, max-age=0, must-revalidate",
      },
    ],
    "location": [
      {
        "key": "Location",
        "value": "",
      },
    ],
    "pragma": [
      {
        "key": "Pragma",
        "value": "no-cache",
      },
    ],
    "set-cookie": [
      {
        "key": "Set-Cookie",
        "value": "cloudfront.accessToken=MOCK_ACCESS_TOKEN; Domain=d20o2ten651kje.cloudfront.net; Path=/; Expires=Mon, 21 Oct 2024 20:28:19 GMT; Secure",
      },
      {
        "key": "Set-Cookie",
        "value": "cloudfront.idToken=MOCK_ID_TOKEN; Domain=d20o2ten651kje.cloudfront.net; Path=/; Expires=Mon, 21 Oct 2024 20:28:19 GMT; Secure",
      },
      {
        "key": "Set-Cookie",
        "value": "cloudfront.refreshToken=MOCK_REFRESH_TOKEN; Domain=d20o2ten651kje.cloudfront.net; Path=/; Expires=Mon, 21 Oct 2024 20:28:19 GMT; Secure",
      },
    ],
  },
  "status": "302",
}
`;
